File: src/main/java/net/cjsah/console/command/argument/StringArgument.java
package net.cjsah.console.command.argument;

import net.cjsah.console.command.StringReader;
import net.cjsah.console.command.context.CommandContext;
import net.cjsah.console.exceptions.CommandException;

public class StringArgument implements Argument<String> {
    private final boolean space;

    private StringArgument(final boolean space) {
        this.space = space;
    }

    public static StringArgument word() {
        return new StringArgument(false);
    }

    public static StringArgument string() {
        return new StringArgument(true);
    }

    public static String getString(CommandContext context, String name) {
        return context.getArgument(name, String.class);
    }

    @Override
    public String parse(StringReader reader) throws CommandException {
        if (this.space) {
            String text = reader.getRemaining();
            reader.setCursor(reader.getTotalLength());
            return text;
        }else return reader.readUnquotedString();
    }
}

File: src/main/kotlin/net/cjsah/console/command/argument/StringArgument.kt
@file:Suppress("unused")

package net.cjsah.console.command.argument

import net.cjsah.console.command.StringReader
import net.cjsah.console.command.context.CommandContext

class StringArgument private constructor(private val space: Boolean): Argument<String> {

    companion object {
        @JvmStatic
        fun word() = StringArgument(false)

        @JvmStatic
        fun string() = StringArgument(true)

        @JvmStatic
        fun getString(context: CommandContext, name: String) = context.getArgument(name, String::class.java)
    }

    override fun parse(reader: StringReader): String {
        return if (space) reader.getRemaining().apply { reader.setCursor(reader.getTotalLength()) }
        else reader.readUnquotedString()
    }

}